cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,33 +1242,103 @@ struct Inequality {
                             /*insert_implicit_casts=*/false, &exec_context);
   }
 
+  /// Simplify an is_in predicate against this inequality as a guarantee.
+  Result<Expression> SimplifyIsIn(Expression expr) {
+    const auto& guarantee = *this;
+    auto call = expr.call();
+    auto options = checked_pointer_cast<SetLookupOptions>(call->options);
+
+    auto value_set = options->value_set.make_array();
+    if (!value_set) return expr;
+    if (value_set->length() == 0) return literal(false);
+
+    if (!options->sorted_and_deduped) return expr;
+
+    // For now, only simplify when the guarantee is non-nullable.
+    if (guarantee.nullable) return expr;
+
+    auto compare = [&value_set, &guarantee](size_t i) -> Result<Comparison::type> {
+      ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> scalar, value_set->GetScalar(i));
+      // Nulls compare greater than any non-null value.
+      if (!scalar->is_valid) {
+        return Comparison::GREATER;
+      }
```

Comment:
Why not `Comparison::NA`? SQL might not set nulls-first? This might affect the case when the expr cannot simplified

Reply:
I commented in `SetLookupOptions::sorted_and_deduped` that nulls are required to be placed at the end. I don't particularly like the interface, though, so open to other suggestions on how to handle nulls in the value set.

If the binary search gets `Comparison::NA`, I suppose we would fail simplification and immediately return `expr`? If we allow arbitrary null placement then this might be weird in that the simplification will sometimes succeed and sometimes fail, depending on whether the binary search happens to check a null value. But I'm fine with this approach if you think this is right.